File: TopCoder/python/LRUCache.py
class Node:
    def __init__(self, key, val):
        self.value = val
        self.key = key
        self.next: Node = None
        self.prev: Node = None


class LRUCache:
    def __init__(self, capacity: int):
        self.max_capacity = capacity
        self.map_entries = {}
        self.head = Node(-1, -1)
        self.tail = self.head
        self.curr = 0

    def get(self, key: int) -> int:
        if key in self.map_entries.keys():
            node: Node = self.map_entries[key]
            self.movetolast(node)
            return node.value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if len(self.map_entries.keys()) == self.max_capacity:
            if key in self.map_entries.keys():
                node: Node = self.map_entries[key]
                self.movetolast(node)
                node.value = value
            else:
                node: Node = self.head.next
                self.movetolast(node)
                del self.map_entries[node.key]
                node.value = value
                node.key = key

        else:
            if key in self.map_entries.keys():
                node: Node = self.map_entries[key]
                node.value = value
                self.movetolast(node)
            else:
                node = Node(key, value)
                self.movetolast(node)

        self.map_entries[key] = self.tail

    def movetolast(self, node: Node):

        if node == self.tail:
            return

        if node.prev:
            node.prev.next = node.next

        if node.next:
            node.next.prev = node.prev

        node.next = None
        node.prev = None

        self.tail.next = node
        node.prev = self.tail

        self.tail = node


if __name__ == "__main__":
    cache = LRUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    print(cache.get(1))
    cache.put(3, 3)
    print(cache.get(2))
    cache.put(4, 4)
    print(cache.get(1))
    print(cache.get(3))
    print(cache.get(4))

File: TopCoder/src/main/java/com/leetcode/LRUCache.java
package com.leetcode;

import java.util.HashMap;
import java.util.Map;

//https://leetcode.com/problems/lru-cache/
public class LRUCache {
    class LRUNode {
        int key;
        int value;
        LRUNode next;
        LRUNode prev;

        public LRUNode(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
    private int maxCapacity;
    private int currentCapacity;
    private Map<Integer, LRUNode> map = new HashMap<>();
    private LRUNode head = new LRUNode(-1, -1);
    private LRUNode tail = head;

    public LRUCache(int capacity) {
        this.maxCapacity = capacity;
        this.currentCapacity = 0;
    }

    public int get(int key) {
        if (map.containsKey(key)) {
            int temp = map.get(key).value;
            LRUNode tempNode = map.get(key);
            moveToLastPosition(tempNode);
            return temp;
        }
        return -1;
    }

    public void put(int key, int value) {
        if (currentCapacity == maxCapacity) {
            if (!map.containsKey(key)) {
                LRUNode tempNode = head.next;
                int oldValue = tempNode.key;
                moveToLastPosition(tempNode);
                tempNode.value = value;
                tempNode.key = key;
                map.remove(oldValue);
            } else {
                // already have the element, just update it
                LRUNode tempNode = map.get(key);
                moveToLastPosition(tempNode);
                tempNode.value = value;
            }
        } else {
            // we are under capacity. just append it.
            if (!map.containsKey(key)) {
                LRUNode newNode = new LRUNode(key, value);
                moveToLastPosition(newNode);
                currentCapacity++;
            } else {
                LRUNode tempNode = map.get(key);
                moveToLastPosition(tempNode);
                tempNode.value = value;
            }

        }
        map.put(key, tail);
    }

    private void moveToLastPosition(LRUNode node) {
        if (node == this.tail) {
            return;
        }
        if (node.prev != null) {
            node.prev.next = node.next;
        }

        if (node.next != null) {
            node.next.prev = node.prev;
        }

        node.next = null;
        node.prev = null;

        tail.next = node;
        node.prev = tail;
        tail = tail.next;
    }


    public static void main(String[] args) {
        /*
        LRUCache cache = new LRUCache( 1);
        System.out.println(cache.get(1));
        cache.put(2, 1);
        System.out.println(cache.get(2));
        cache.put(3, 2);
        System.out.println(cache.get(2));
        System.out.println(cache.get(3));
         */
        LRUCache cache = new LRUCache( 10);
        cache.put(10, 13);
        cache.put(3, 17);
        cache.put(6, 11);
        cache.put(10, 5);
        cache.put(9, 10);
        System.out.println(cache.get(13));
        cache.put(2, 19);
        System.out.println(cache.get(2));
        System.out.println(cache.get(3));
        cache.put(5, 25);
        System.out.println(cache.get(8));
        cache.put(9, 22);
        cache.put(5, 5);
        cache.put(1, 30);
        System.out.println(cache.get(11));
        cache.put(9, 12);
        System.out.println(cache.get(7));
        System.out.println(cache.get(5));
        System.out.println(cache.get(8));
        System.out.println(cache.get(9));
        cache.put(4, 30);
        cache.put(9, 3);
        System.out.println(cache.get(9));
        System.out.println(cache.get(10));
        System.out.println(cache.get(10));
        cache.put(6, 14);
        cache.put(3, 1);
        System.out.println(cache.get(3));
        cache.put(10, 11);
        System.out.println(cache.get(8));
        cache.put(2, 14);
        System.out.println(cache.get(1));
        System.out.println(cache.get(5));
        System.out.println(cache.get(4));
        cache.put(11, 4);
        cache.put(12, 24);
        cache.put(5, 18);


    }
}
